File: src/03.number/31.50.go
package number_bit

func MyPow(x float64, n int) float64 {
	if x == 0.0 {
		return 0.0
	}

	b := int64(n)
	res := 1.0

	if b < 0 {
		b = -b
		x = 1 / x
	}

	for b > 0 {
		// b 如果是奇数
		if (b & 1) == 1 {
			res *= x
		}
		x *= x
		b >>= 1
	}

	return res
}

File: src/03.number/31.50.ts
export function myPow(x: number, n: number): number {
    // if (x === 0) {
    //     return 0.0;
    // }

    // let res = 1.0;
    // if (n < 0) {
    //     x = 1 / x;
    //     n = -n;
    // }

    // while (n > 0) {
    //     if ((n & 1) === 1) {
    //         res *= x;
    //     }
    //     x *= x;
    //     n >>>= 1;
    // }

    // return Number(res.toFixed(5));

    // 高效求幂
    // a^b = ?
    // b 为奇数时, a^b = a * a^(b-1)
    // b 为偶数时, (a^(b/2))^2

    if (n === 0 || n === 1) {
        return n === 0 ? 1 : x;
    } else if (n < 0) {
        return myPow(1 / x, -n);
    } else {
        return n & 1 ? myPow(x * x, n >>> 1) * x : myPow(x * x, n >>> 1);
    }
}
